ConcernsCaseWork/ConcernsCaseWork/Pages/Case/Management/Action/Nti/AddConditions.cshtml.cs

```diff
@@ -0,0 +1,173 @@
+			INtiWarningLetterReasonsCachedService ntiWarningLetterReasonsCachedService,
+			INtiModelService ntiModelService,
+			INtiConditionsCachedService ntiConditionsCachedService,
+			ILogger<NtiWarningLetterConditionModel> logger)
+		{
+			_ntiWarningLetterStatusesCachedService = ntiWarningLetterStatusesCachedService;
+			_ntiWarningLetterReasonsCachedService = ntiWarningLetterReasonsCachedService;
+			_ntiModelService = ntiModelService;
+			_ntiConditionsCachedService = ntiConditionsCachedService;
+			_logger = logger;
+		}
+
+		public async Task<IActionResult> OnGetAsync()
+		{
+			_logger.LogInformation("Case::Action::NTI-WL::AddConditionsPageModel::OnGetAsync");
+
+			try
+			{
+				if (ContinuationId == null)
+				{
+					throw new InvalidOperationException("Continuation Id not found");
+				}
+
+				ExtractCaseUrnFromRoute();
+				ExtractWarningLetterIdFromRoute();
+
+				var model = await GetUpToDateModel();
+				SelectedConditions = model.Conditions;
+
+				AllConditions = await _ntiConditionsCachedService.GetAllConditionsAsync();
+
+				IsReturningFromConditions = true;
+				TempData.Keep(nameof(ContinuationId));
+				return Page();
+			}
+			catch (Exception ex)
+			{
+				_logger.LogError("Case::NTI-WL::AddConditionsPageModel::OnGetAsync::Exception - {Message}", ex.Message);
+
+				TempData["Error.Message"] = ErrorOnGetPage;
+				return Page();
+			}
+		}
+
+		public async Task<IActionResult> OnPostAsync(string action)
+		{
+			try
+			{
+				ExtractCaseUrnFromRoute();
+				ExtractWarningLetterIdFromRoute();
+
+				AllConditions = await _ntiConditionsCachedService.GetAllConditionsAsync();
+				var conditions = Request.Form["condition"];
+				var model = await GetUpToDateModel();
+				model.Conditions = conditions.Select(s => NtiMappers.ToServiceModel(AllConditions.Single(c => c.Id == int.Parse(s)))).ToArray();
+				await _ntiModelService.StoreNti(model, ContinuationId);
+
+				IsReturningFromConditions = true;
+				TempData.Keep(nameof(ContinuationId));
+
+				if (NtiId != null)
+				{
+					return Redirect($"/case/{CaseUrn}/management/action/nti/{NtiId}/edit");
+				}
+				else
+				{
+					return Redirect($"/case/{CaseUrn}/management/action/nti/add");
+				}
+			}
+			catch (Exception ex)
+			{
+				_logger.LogError("Case::NTI::AddConditionsPageModel::OnPostAsync::Exception - {Message}", ex.Message);
+				TempData["Error.Message"] = ErrorOnPostPage;
+			}
+
+			return Page();
+		}
+
+		public bool IsConditionSelected(NtiConditionDto condition)
+		{
+			return SelectedConditions?.Any(c => c.Id == condition.Id) ?? false;
+		}
+
+		private async Task<NtiModel> GetUpToDateModel()
+		{
+			return await _ntiModelService.GetNtiAsync(ContinuationId);
+		}
+
+		private void ExtractCaseUrnFromRoute()
+		{
+			if (TryGetRouteValueInt64("urn", out var caseUrn))
+			{
+				CaseUrn = caseUrn;
+			}
+			else
+			{
+				throw new InvalidOperationException("CaseUrn not found in the route");
```

Comment:
I've seen both InvalidOperationException and Exception thrown when failing to get CaseUrn from route. Is InvalidOperationException the preferred going forward?

Reply:
Yes, InvalidOperationException is the most suitable exception to be thrown here.